### PR TITLE
test/main: refactor process-wide setup/shutdown.

### LINKED ISSUE #6689: ares_library_cleanup() won't be called if an exception occurs in MainCommonBase constructor.
*Title*: ares_library_cleanup() won't be called if an exception occurs in MainCommonBase constructor.

*Description*:
The MainCommonBase constructor sets up some global state; specifically, it calls ares_library_init() as its first line.  That global state is reset in ~MainCommonBase via ares_library_cleanup().   If an exception is thrown from a function called from the MainCommonBase constructor (as happens if there is an envoy configuration error), the init function will be called by the cleanup function won't be.  That may be problematic for tests that use MainCommonBase which may be testing invalid envoy configs--there will be global state droppings in place after a construction attempt that throws.

(I don't know the ares library so I don't now how important this is--I ran into this tracing a similar problem with the thread singleton factory, but I expect that problem to be fixed by pull request #6658 .]

https://github.com/envoyproxy/envoy/blob/master/source/exe/main_common.cc#L53
https://github.com/envoyproxy/envoy/blob/master/source/exe/main_common.cc#L101

@jmarantz

### LINKED ISSUE #6839: tsan caught ares cleanup shutdown bug
https://pastebin.com/FKsvL1sM

from

https://circleci.com/gh/envoyproxy/envoy/211476?utm_campaign=workflow-failed&utm_medium=email&utm_source=notification

tl;dr looks like both google's grpc code and DnsResolverImpl want to do ares cleanup.

Not super high pri as it's a shutdown bug but worth fixing if for no other reason to keep CI happy

### PR DESCRIPTION
**Pull request by htuch**
We have quite a few third party libraries that need to have process wide
setup/shutdown (including nghttp2, c-ares, gRPC, libevent). Ideally we
have a single source of truth for this sequence.

This PR is motivated by a TSAN race in CI reported in #6839, where the
DnsResolverImpl was doing its own c-ares setup/shutdown, racing with
gRPC library.

While we're mucking around in this code, also fix #6689.

Risk level: Low
Testing: bazel test --config=clang-tsan -c dbg
  //test/server:server_fuzz_test --runs_per_test=5000

Fixes #6839
Fixes #6689

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/exe/process_wide.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "common/common/cleanup.h"
+
+namespace Envoy {
+
+// Process-wide lifecycle events for global state in third-party dependencies,
+// e.g. gRPC, c-ares.
+class ProcessWide {
```
Comment (mattklein123) [inline source/exe/process_wide.h]: Very mild preference to just have the `ProcessWide` constructor do the setup, and then it's destructor do the cleanup, but don't feel strongly about it if you prefer it this way.

File: source/exe/process_wide.cc [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#include "ares.h"
+
+#ifdef ENVOY_GOOGLE_GRPC
+#include "grpc/grpc.h"
+#endif
+
+namespace Envoy {
+
+CleanupPtr ProcessWide::setup() {
```
Comment (curiouserrandy) [inline source/exe/process_wide.cc]: Would it be worthwhile to have some sort of check that this function is only called once, or (preferably) that calls to it are matched with interleaved cleanup calls (for testing contexts)?

File: source/exe/process_wide.h (hunk 1 above)
Comment (curiouserrandy) [reply source/exe/process_wide.h]: +1

File: source/exe/process_wide.cc [hunk 3]
```diff
@@ -13,8 +13,14 @@
 #endif
 
 namespace Envoy {
+namespace {
+// Static variable to sanity check init state.
+bool process_wide_iniitalized;
```
Comment (alyssawilk) [inline source/exe/process_wide.cc]: iniitalized -> initialized

File: source/exe/process_wide.cc [hunk 4]
```diff
@@ -13,8 +13,14 @@
...
+} // namespace
 
 ProcessWide::ProcessWide() {
+  ASSERT(!process_wide_iniitalized);
```
Comment (alyssawilk) [inline source/exe/process_wide.cc]: isn't process_wide_iniitalized uninitialized at this point?

Comment (htuch) [reply source/exe/process_wide.cc]: It's a static variable, so gets zero init. It's now a `uint32_t` as well, due to how `main_common_test` works.
